[4단계] 로또(수동)

RoadtoS7 opened the pull request:
![스크린샷 2022-05-22 오후 4 01 39](https://user-images.githubusercontent.com/57262833/169683066-00d7c566-da62-4746-86ce-f66ddd64e413.jpg)

리뷰어님 피드백 감사합니다! 이번에도 잘 부탁드립니다!!~💃
1. 힌트에서 주어진진대로 Lotto 숫자 한개를 `LottoNumber` 라는 구조체에 담아보았습니다!
2. LottoSeller에서 `주어진 금액으로 살 수 있는 로또 개수 계산`,`LottoBag인스턴스 생성` 이렇게 2개의 책임을 가지고 있었는데, LottoBagFactory에게 `LottoBag 생성 책임`을 주고, LottoSeller는 `로도 개수 계산` 책임만 갖도록 수정했습니다!
3. 힌트에 서술된대로 제가 논리적 오류에만 error를 던지도록 적절히 구현했는지 궁금합니다!    
(제 프로그램에서 논리적 오류가 어떤 것에 해당하는지 고민해보았는데, 거의 대부분이 논리적 오류인 것 같아서.. 대부분의 경우에 throw를 하고 있더라구요 😂 )

RoadtoS7 committed:
refactor: InputView의 인스턴스를 생성하지 못하도록 enum으로 수정

InputView는 사용자의 입력을 받는 static 메서드로 구성되어있으므로 인스턴스를 생성할 필요가 없다
따라서 인스턴스 생성을 막기 위해서 enum으로 타입을 변경했다

RoadtoS7 committed:
feat: InputView에서 수동 로또 개수를 입력받을 수 있도록 수정

RoadtoS7 committed:
feat: InputView에서 수동 로또 번호를 입력받을 수 있도록 수정

RoadtoS7 committed:
feat: CustomLotto intializer 테스트 케이스 추가

RoadtoS7 committed:
feat: CustomLotto 테스트 케이스 추가하는 intializer 로직 구현

RoadtoS7 committed:
feat: LottoNumberValidatorTest에 LottoNumber의 개수를 체크하는 유닛테스트케이스 추가

RoadtoS7 committed:
refactor: LottoNumberValidator.validate 리팩터링

RoadtoS7 committed:
refactor: LottoNumberValidator에서 validate를 로또 숫자 범위와 로또 숫자 개수를 체크하는 메서드로 분리

RoadtoS7 committed:
refactor: CustomLotto, UserLotto, WinningLotto에서 수정한 LottoValidator 사용하도록 수정

RoadtoS7 committed:
feat: LottoParser에 CustomLotto를 parse하는 기능 추가

RoadtoS7 committed:
feat: LottoSeller.sellLotto가 수동 로또를 받을 수 있도록 수정 및 테스트 케이스 구현

RoadtoS7 committed:
refactor: LottoBat이 UserLotto가 아니라 Lotto를 갖도록 수정, LottoSeller.sellLotto 동작 구현

RoadtoS7 committed:
refactor: LottoSeller.selLotto 동작 리팩터링

RoadtoS7 committed:
refactor: main에서 변수 lottoTicket 이름 -> lottoBag으로 수정

RoadtoS7 committed:
refactor: LottoTicketFormatter 이름 -> LottoBagFormatter로 수정

RoadtoS7 committed:
refactor: LottoBagForammter에서 로또 장수도 출력하도록 수정

RoadtoS7 committed:
feat: main에서 수동 로또 번호 입력받도록 구현

RoadtoS7 committed:
feat: LottoCount에 대한 유닛 테스트 케이스 추가

RoadtoS7 committed:
feat: 로또 개수를 나타내는 LottoCount 구현

RoadtoS7 committed:
fix: WinningRecord.calculateProfite 로직 변경에 따라 테스트코드도 수정

RoadtoS7 committed:
fix: 번호 3개 당첨된 로또 번호 개수 출력되지 않는 이슈 해결

RoadtoS7 committed:
feat: LottoParser 및 유닛테스트 구현

RoadtoS7 committed:
feat: InputView에 LottoCount이용하여 수동 로또 번호 입력받는 메서드 추가

RoadtoS7 committed:
refactor: main에서 로또 숫자 개수를 LottoCount로 다루도록 수정

RoadtoS7 committed:
refactor: OutputView에서 Lotto개수 출력하는 메서드 제거

LottoBag을 출력할 때 같이 출력하도록 수정했기에, Lotto개수만 출력하는 메서드는 제거한다

RoadtoS7 committed:
refactor: LottoBag, LottoBagFormatter에서 LottoCount 사용하도록 수정

RoadtoS7 committed:
refactor: LottoBag에서 총 lotto 개수를 알려주는 lottoCount프로퍼티 제거

RoadtoS7 committed:
feat: 원시값 포장을 위해 LottoNumber 구현

RoadtoS7 committed:
feat: UserLotto, WinningLotto, CustomLotto에서 LottoNumber를 사용하도록 수정

RoadtoS7 committed:
refactor: 사용자 안내문이나 결과물을 출력할 때 줄바꿈이 일어나도록 수정

RoadtoS7 committed:
refactor: LottoBagFactory에서 수동 로또를 담고 있는 LottoBag을 생성할 수 있도록 수정

RoadtoS7 committed:
refactor: LottoSeller가 LottoBagFacotry를 이용하여 LottoBag을 생성하도록 수정

- 수정 이유
LottoSeller는 주어진 돈으로 살 수 있는 로또 개수를 계산하는 역할만 수행하도록 하기 위해,
LottoBag을 생성하는 역할은 LottoBagFactory에게 위임한다.

RoadtoS7 committed:
refactor: 모든 로또는 생성될 때 자신만의 LottoValidator들을 가지고 로또 숫자 규칙을 만족하는지 체크하도록 수정

- 수정 이유
모든 로또가 LottoNumberValidator를 이용하여 로또 규칙을 만족하는지 체크하는 것은 LottoNumberValidator에 대한 결합도가 높은 로직이라고 생각했다.
만일 다른 규칙을 갖는 로또가 등장할 경우, LottoNumberValidator에 해당 규칙을 점검하는 메서드를 생성해주어야 한다.
따라서 이런 상황을 예방하기 위해 로또 스스로가 자신이 어떤 규칙으로 로또번호를 보관하는지 점검하도록 수정했다.

RoadtoS7 commented on Lotto/Lotto/Lotto/WinningLotto.swift:
```diff
@@ -7,16 +7,20 @@
-       
         let lottoNumbers = numbers.map { LottoNumber($0) }
         self.numbers = Set(lottoNumbers)
         self.bonusNumber = bonusNumber
```
LottoNumberValidator가 로또 번호들이 규칙(6개여야 한다. 1~45 이내의 숫자여야 한다.)을 만족하는지를 전담해서 체크하고 있는데, 모든 로또 (UserLotto, CustomLotto, WinnigLotto)와 LottoNumberValidator 간의 결합도가 높다고 판단했습니다!

따라서 `LottoNumberValidator`에게 로또 번호 규칙을 만족하는지 묻는 것이 아니라,  `LottoValidator` 구현체 array를 각 Lotto가 가지고서 스스로 체크할 수 있도록 수정했습니다.

_그런데 WinningLotto의 경우 bonus 숫자를 추가로 가지는데, 이 bonus 숫자는 WinningLotto의 모든 `WinningValidator`의 체크를 받는 것이 아니라, 일부 `WiningValidator`의 체크만 받아야 합니다. 따라서  중간에 if문을 통해 bonus number에도 적용해야하는 LottoValidator를 찾는 로직이 들어가게 되었습니다..!_


이 부분을 개선하고 싶은데 마땅한 방법이 떠오르지 않습니다... 이 방법이 최선일까요..🥲

wonhee009 replied:
보너스를 제외한 로또 번호의 검증이 모두 끝난 후 보너스 검증을 해야하는거로 이해했습니다.
Array에 모든 validator가 정의되어 있어 생긴 문제라고 생각합니다. 현재는 filter를 걸 validator가 Range 하나이지만 언제든지 다른 filter문이 추가될 수 있다고 생각합니다.

저는 Validator도 유틸성이 높지 않나 생각해 아래와 같이 static 메서드를 사용해도 문제 없다고 생각합니다.
```swift
struct Validator {
   static func range(lotto: Lotto) { ... }
   static func count(lotto: Lotto) { ... }
}
```

아니면 Lotto를 하나의 super class로 두고 그 안에 validte를 두는 것도 괜찮을거 같구요.

RoadtoS7 replied:
저번 step3에서 LottoNumberValidator라는 유틸 클래스로 구현했어서, 이번에는 피드백 주신 대로 Lotto를 super class로 두고 template method 패턴을 사용해보았습니다!
[cf22926](https://github.com/next-step/swift-lotto/pull/31/commits/cf229267862d01575aa98ef3b3bd0e0c9f64d6af) 커밋에서 반영했습니다! 피드백 감사합니다!!~

wonhee009 replied:
`CustomLotto`, `UserLotto`가 `Lotto`와 다른 점이 없어보이는데요. 
`Lotto`에서 `validate` 메서드를 갖기 보다는 `validateRange`, `validateCount`만 가지고 있고 각각의 Lotto 타입들(CustomLotto, UserLotto, WinningLotto)에서 원하는 validate를 호출하는건 어떨까요?
현재 구조에서는 WinningLotto가 어떤걸 validate하는지 한 눈에 보이는 반면, CustomLotto, UserLotto가 어디부터 어디까지를 Validate하고 있는지 파악이 어려울거 같습니다.

RoadtoS7 replied:
> CustomLotto, UserLotto가 Lotto와 다른 점이 없어보이는데요.

UserLotto는 자동로또로서 사용하고, CustomLotto는 수동로또로 사용하기 위해 둘을 따로 구현했었습니다..!
UserLotto라는 이름이 자동로또로 인식되지 않는 것 같아서, AutoLotto로 이름을 수정했습니다! ([1d7cbb5](https://github.com/next-step/swift-lotto/pull/31/commits/1d7cbb5c332e8487d27a4d4256c231365d9e33f5) 커밋)

> 현재 구조에서는 WinningLotto가 어떤걸 validate하는지 한 눈에 보이는 반면, CustomLotto, UserLotto가 어디부터 어디까지를 Validate하고 있는지 파악이 어려울거 같습니다.

파악이 어려울 수 있겠네요.! 피드백 감사합니다!!
[76179e7](https://github.com/next-step/swift-lotto/pull/31/commits/76179e7523c45dc2447872fefebb5c3298ab4888) 커밋에서 반영했는데, 이렇게 구현하는 것을 의미하신 게 맞는지 궁금합니다!

wonhee009 commented on Lotto/Lotto/Lotto.swift:
```diff
@@ -15,13 +15,13 @@ enum LottoConstant {
     func contains(number: Int) -> Bool {
-        numbers.contains(number)
+        numbers.contains(LottoNumber(number))
     }
```
```swift
func contains(number: LottoNumber) -> Bol { ... } 
```
`LottoNumber`로 통일성을 가져갈 필요는 없을까요?

RoadtoS7 replied:
[99d0c11](https://github.com/next-step/swift-lotto/pull/31/commits/99d0c115d3a37cd98fd3c655ab718dfac73583ed) 커밋에서 반영했습니다! 피드백 감사합니다!!

wonhee009 commented on Lotto/Lotto/Lotto/LottoNumberValidator.swift:
```diff
@@ -7,11 +7,19 @@
-    static func validate<LottoNumbers: Collection>(numbers: LottoNumbers) -> Bool where LottoNumbers.Element == Int {
-        
-        return numbers.reduce(true) { partialResult, number in
+struct LottoNumberValidator<LottoNumbers: Collection> where LottoNumbers.Element == Int {
+    
+    static func validateNumberCount(of numbers: LottoNumbers) throws {
+        let setNumbers = Set(numbers)
+        guard setNumbers.count == LottoConstant.numberCount else {
+            throw LottoError.invalidNumberCount(setNumbers.count)
+        }
+    }
+    
+    static func validateNumberRange(of numbers: LottoNumbers) throws {
+        let result = numbers.reduce(true) { partialResult, number in
             partialResult && LottoConstant.numberRange.contains(number)
         }
+        guard result else { throw LottoError.invalidLottoNumber }
     }
 }
```
`CountValidator`, `RangeValidator` struct로 검증해주고 있는거로 이해했는데요.
해당 메서드가 여전히 남아 있는 이유는 테스트 때문일까요? (현재 사용처가 테스트 코드에만 있는거 같아 질문드립니다.)
만약 테스트를 위해 남아 있는 코드라면 구조를 다시 생각해보는건 어떨까요?

RoadtoS7 replied:
더이상 사용하지 않아서 제거했어요! 
[9f52c9a](https://github.com/next-step/swift-lotto/pull/31/commits/9f52c9ae0d784c6cced87ebae790df9b32470d9f) 
 커밋에서 반영했습니다! 피드백 감사합니다!

wonhee009 commented on Lotto/Lotto/Lotto/WinningLottoRule.swift:
```diff
@@ -0,0 +1,8 @@
+//
+//  WinningLottoRule.swift
+//  Lotto
+//
+//  Created by nylah.j on 2022/05/19.
+//
+
+import Foundation
```
사용하지 않는 파일일까요?

RoadtoS7 replied:
엇 사용하지 않는 파일이어서 제거했습니다! 피드백 감사합니다!!

wonhee009 commented on Lotto/Lotto/LottoCountParser.swift:
```diff
@@ -0,0 +1,29 @@
+
+import Foundation
+
+enum LottoCountParser {
+    enum Error: LocalizedError {
+        case invalid
+        
+        var errorDescription: String? {
+            switch self {
+            case .invalid: return "로또 개수는 0개 이상이어야 합니다."
+            }
+        }
+    }
+    
+    static func parse(_ lottoCountInput: String?) throws -> LottoCount {
+        guard let input = lottoCountInput,
+              let lottoCountValue = Int(input),
+              let lottoCount = LottoCount(lottoCountValue) else {
+                  throw Error.invalid
+              }
+        return lottoCount
+    }
+}
```
parser를 enum으로 정의한 이유가 있을까요?

RoadtoS7 replied:
static 함수만 존재하는 util성 클래스/구조체라서, 인스턴스를 만들어서 사용하지 못하도록 하기 위해 enum을 붙여주었습니다..!
이렇게 했을 때, util성 클래스/구조체의 인스턴스가 불필요하게 생성되는 일을 막을 수 있다고 생각했는데 혹시 이 방법에서 제가 생각치 못한 문제점이 있을까요..?

wonhee009 replied:
> util성 클래스/구조체의 인스턴스가 불필요하게 생성되는 일을 막을 수 있다고 생각했는데

저도 이렇게 생각하는데요. 이렇게 기능적 Util에는 enum보다 struct를 사용한 경우를 더 많이 본거 같아서 이유가 있나 질문 드렸습니다! :)

RoadtoS7 replied:
보통 Util 에는 enum 보다 struct를 더 많이 사용하나요??

wonhee009 commented on Lotto/Lotto/Util/WinningRecordFormatter.swift:
```diff
@@ -18,7 +18,7 @@ struct WinningRecordFormatter {
     
     static func format(_ statistics: WinningRecord) -> String {
-        let fifth = String(format: Text.fifth, statistics.value[.first]!) + newLine
+        let fifth = String(format: Text.fifth, statistics.value[.fifth]!) + newLine
```
👍

wonhee009 reviewed (changes requested):
수고하셨습니다 🎉 
코멘트 확인 부탁드려요 🙏

RoadtoS7 committed:
refactor: LottoNubmer에서 contains함수의 파라미터로 Int -> LottoNumber 사용하도록 수정

RoadtoS7 committed:
refactor: LottoNumberValidator 제거

- 제거 이유
더이상 사용하지 않는다

RoadtoS7 commented:
힌트에 서술된대로 제가 논리적 오류에만 error를 던지도록 적절히 구현했는지 궁금합니다!
(제 프로그램에서 논리적 오류가 어떤 것에 해당하는지 고민해보았는데, 거의 대부분이 논리적 오류인 것 같아서.. 대부분의 경우에 throw를 하고 있더라구요 😂  논리적 오류를 제가 잘 못 이해하고 있는 것 일까요..?ㅜㅠㅜ)

RoadtoS7 committed:
refactor: Lotto를 super class로 만들고 CustomLotto, UserLotto, WinningLotto가 Lotto를 상속받도록 수정

- 수정 이유
모든 Lotto(CustomLotto, UserLotto, WinningLotto)의 인스턴스가 생성될 때 인자로 받은 로또 숫자 검증 -> 검증 통과한 로또 번호를 가지고 인스턴스 생성, 이 순서로 진행된다.
따라서 템플릿 메서드 패턴으로 super class에서 이 로직의 순서를 정해두고, 필요시에만 각 단계의 메서르르 오버라이딩하여 수정하도록 함

RoadtoS7 committed:
refactor: UserLotto, CustomLotto, WinningLotto에서 Lotto의 validate 메서드 중 필요한 것을 호출하도록 수정

wonhee009 commented on Lotto/Lotto/Lotto.swift:
```diff
@@ -14,17 +14,40 @@ enum LottoConstant {
+class Lotto {
+    var numbers: Set<LottoNumber>
+    
+    init<LottoNumbers>(numbers: LottoNumbers) throws where LottoNumbers: Collection, LottoNumbers.Element == Int {
+        let lottoNumbers = numbers.map { LottoNumber($0) }
+        self.numbers = Set(lottoNumbers)
+        try vaidate(numbers: numbers)
+    }
```
validate는 검증하는 과정이니까 논리적 오류라고 생각하지만 `Lotto.init`에서 검증에 실패하는 것도 논리적 오류일까요?
여기서는 nil을 반환해야하지 않을까 하는게 저의 생각입니다 :)

RoadtoS7 replied:
`Lotto.init` 에서 검증에 실패하는 것이 논리적 오류라고 생각하신 이유를 한번더 여쭙고 싶습니다..! 
`Lotto.init` 에서 검증에 실패하면, `Lotto` 를 생성할 수 없는 상황이고 만일 이때 Lotto를 생성하면 프로그램이 예측한대로 동작하지 않으니까 논리적 오류라고 생각했었습니다..!

wonhee009 replied:
validate까지가 논리적 오류라고 생각했고 Lotto.init 시에 validate로 "인해" 실해하는 케이스는 논리적 오류가 아니라고 생각했습니다.
해당 케이스는 validate라는 논리적 오류에서 따라오는? 실패라고 생각해서 여기서는 nil을 반환하는게 좀 더 맞지 않나 생각했습니다.

RoadtoS7 replied:
[133b282](https://github.com/next-step/swift-lotto/pull/31/commits/133b2826b62b82df8d542f37ea89d21609e39415) 커밋에서 반영했습니다! 
피드백 감사합니다!!

RoadtoS7 committed:
refactor: UserLotto를 AutoLotto로 네이밍 변경

- 수정 이유
UserLotto와 CustomLottor간의 구분이 잘 되지 않으므로, AutoLotto, CustomLotto로 구분하여 만든다

RoadtoS7 committed:
refactor: Lotto생성시 에러가 발생할 경우 error throw하지 않고 nil을 반환하도록 수정

- 수정 이유
validate 메서드는 Lotto를 생성할 대 주어진 숫자들이 로도가 될 수 있는지 체크한다.
체크 과정에서 로또 숫자가 될 수 없다면 이는 논리적 에러의 사유가 되기때문에 error를 throw한다.

하지만 Lotto 이니셜라이저에서 validate 메서드를 호출한 결과로 error를 받는결과는 논리적 오류로 인한 결과이기에, nil을 반환하는 것으로 수정했다.

wonhee009 reviewed (approved):
수고하셨습니다! 🎉
